src/api/networking.rs

```diff
@@ -854,3 +1023,615 @@ fn socket_address(
+            .udp_sockets
+            .get(socket_id)
+            .or_trap("lunatic::network::udp_receive_from")?;
+
+        // Check for timeout first
+        if let Some(result) = tokio::select! {
+            _ = async_std::task::sleep(Duration::from_millis(timeout as u64)), if timeout != 0 => None,
+            result = socket.recv_from(buffer) => Some(result)
+        } {
+            let (opaque, socket_result, return_) = match result {
+                Ok((bytes, socket)) => (bytes as u64, Some(socket), 0),
+                Err(error) => (caller.data_mut().errors.add(error.into()), None, 1),
+            };
+
+            let memory = get_memory(&mut caller)?;
+            memory
+                .write(&mut caller, opaque_ptr as usize, &opaque.to_le_bytes())
+                .or_trap("lunatic::networking::udp_receive_from")?;
+
+            if let Some(socket_addr) = socket_result {
+                let dns_iter_id = caller
+                    .data_mut()
+                    .resources
+                    .dns_iterators
+                    .add(DnsIterator::new(vec![socket_addr].into_iter()));
+                memory
+                    .write(
+                        &mut caller,
+                        dns_iter_ptr as usize,
+                        &dns_iter_id.to_le_bytes(),
+                    )
+                    .or_trap("lunatic::networking::udp_receive_from")?;
+            }
+            Ok(return_)
+        } else {
+            // Call timed out
+            Ok(9027)
+        }
+    })
+}
+
+//% lunatic::networking::udp_connect(
+//%     addr_type: u32,
+//%     addr_u8_ptr: u32,
+//%     port: u32,
+//%     flow_info: u32,
+//%     scope_id: u32,
+//%     timeout: u32,
+//%     id_u64_ptr: u32,
+//% ) -> u32
+//%
+//% Returns:
+//% * 0 on success    - The ID of the newly created UDP listener is written to **id_ptr**.
+//% * 1 on error      - The error ID is written to **id_ptr**
+//% * 9027 on timeout - The socket connect operation timed out.
+//%
+//% Traps:
+//% * If **addr_type** is neither 4 or 6.
+//% * If **addr_u8_ptr** is outside the memory
+//% * If **id_u64_ptr** is outside the memory.
+#[allow(clippy::too_many_arguments)]
+fn udp_connect(
+    mut caller: Caller<ProcessState>,
+    addr_type: u32,
+    addr_u8_ptr: u32,
+    port: u32,
+    flow_info: u32,
+    scope_id: u32,
+    timeout: u32,
+    id_u64_ptr: u32,
+) -> Box<dyn Future<Output = Result<u32, Trap>> + Send + '_> {
+    Box::new(async move {
+        let memory = get_memory(&mut caller)?;
+        let socket_addr = socket_address(
+            &caller,
+            &memory,
+            addr_type,
+            addr_u8_ptr,
+            port,
+            flow_info,
+            scope_id,
+        )?;
+
+        if let Some(result) = tokio::select! {
+            _ = async_std::task::sleep(Duration::from_millis(timeout as u64)), if timeout != 0 => None,
+            result = UdpSocket::bind("127.0.0.1:0") => Some(result)
```

Comment:
I believe the `connect` function is meant to be used on an existing `UdpSocket`, and we shouldn't create a new on `connect`.

It sets an existing `UdpSocket` to filter out any packets coming from a certain source. E.g. I bind to `localhost:4000`, get a socket back, connect this socket to `x.x.x.x:3000`. Now I listen to all incoming messages on `localhost:4000` that are comming from `x.x.x.x:3000`

If you assign every time a random address on `connect` you lose the possibility to control your local address.

Reply:
Thanks for catching this. I misunderstood how it was supposed to work, and just followed the examples I found on the UdpSocket docs.